test/SubmitUpgradeProposal.t.sol

```diff
@@ -0,0 +1,146 @@
+    proxyTreasuryGovernorDeployer = new DeployTreasuryGovernor();
+    proxyCoreGovernorDeployer.setUp();
+    proxyTreasuryGovernorDeployer.setUp();
+
+    // Deploy Governor proxy contracts
+    newCoreGovernor = proxyCoreGovernorDeployer.run(_implementation);
+    newTreasuryGovernor = proxyTreasuryGovernorDeployer.run(_implementation);
+
+    // Current governors and timelocks
+    currentCoreGovernor = GovernorUpgradeable(payable(ARBITRUM_CORE_GOVERNOR));
+    currentCoreTimelock = TimelockControllerUpgradeable(payable(ARBITRUM_CORE_GOVERNOR_TIMELOCK));
+    currentTreasuryGovernor = GovernorUpgradeable(payable(ARBITRUM_TREASURY_GOVERNOR));
+    currentTreasuryTimelock = TimelockControllerUpgradeable(payable(ARBITRUM_TREASURY_GOVERNOR_TIMELOCK));
+
+    // Deploy a mock ArbSys contract at ARB_SYS
+    MockArbSys mockArbSys = new MockArbSys();
+    bytes memory code = address(mockArbSys).code;
+    vm.etch(ARB_SYS, code);
+  }
+
+  function test_SuccessfullyExecuteUpgradeProposal() public {
+    // Propose
+    (
+      address[] memory _targets,
+      uint256[] memory _values,
+      bytes[] memory _calldatas,
+      string memory _description,
+      uint256 _proposalId
+    ) = submitUpgradeProposal.run(
+      ARBITRUM_CORE_GOVERNOR //maybe also the treasury governor to use in createProposal
+    );
+    assertEq(uint256(currentCoreGovernor.state(_proposalId)), uint256(IGovernor.ProposalState.Pending));
+    vm.roll(vm.getBlockNumber() + currentCoreGovernor.votingDelay() + 1);
+    assertEq(uint256(currentCoreGovernor.state(_proposalId)), uint256(IGovernor.ProposalState.Active));
+
+    // Vote
+    for (uint256 i; i < _majorDelegates.length; i++) {
+      vm.prank(_majorDelegates[i]);
+      currentCoreGovernor.castVote(_proposalId, uint8(VoteType.For));
+    }
+
+    // Success
+    vm.roll(vm.getBlockNumber() + currentCoreGovernor.votingPeriod() + 1);
+    assertEq(uint256(currentCoreGovernor.state(_proposalId)), uint256(IGovernor.ProposalState.Succeeded));
+
+    // Queue
+    currentCoreGovernor.queue(_targets, _values, _calldatas, keccak256(bytes(_description)));
+    assertEq(uint256(currentCoreGovernor.state(_proposalId)), uint256(IGovernor.ProposalState.Queued));
+    vm.warp(vm.getBlockTimestamp() + currentCoreTimelock.getMinDelay() + 1);
+
+    // Execute
+    currentCoreGovernor.execute(_targets, _values, _calldatas, keccak256(bytes(_description)));
+    assertEq(uint256(currentCoreGovernor.state(_proposalId)), uint256(IGovernor.ProposalState.Executed));
+  }
+
+  function test_ExecuteUpgradeUsingUpgradeExecutor() public {
+    TimelockRolesUpgrader timelockRolesUpgrader = new TimelockRolesUpgrader(
+      ARBITRUM_CORE_GOVERNOR_TIMELOCK,
+      ARBITRUM_CORE_GOVERNOR,
+      address(newCoreGovernor),
+      ARBITRUM_TREASURY_GOVERNOR_TIMELOCK,
+      ARBITRUM_TREASURY_GOVERNOR,
+      address(newTreasuryGovernor)
+    );
+
+    address target = address(timelockRolesUpgrader);
```

Comment:
Is the thought here that we can't get the proposal from `ArbSys` to `UpgradeExecutor`, so we just test that the upgrade contract does what we expect? I think it would be nice to explain in natspec why we have this issue, and what tradeoffs we're making in the tests as a result. I also wonder if we can etch over ArbSys to unwrap the calldata and pass the last bit to UpgradeExecutor, because currently, we don't have much guarantee that the script in SubmitUpgradeProposal.s.sol is actually creating the calldata we need. I'd be fine to move that to a part 2 PR, let me know what you think on this.

Reply:
we are going to follow up on this in the next PR!